source/common/protobuf/utility.cc

```diff
@@ -586,6 +590,148 @@ std::string MessageUtil::CodeEnumToString(ProtobufUtil::error::Code code) {
+  auto* any = dynamic_cast<ProtobufWkt::Any*>(message);
+  if (any != nullptr) {
+    Protobuf::DynamicMessageFactory message_factory;
+    auto typed_message = typeUrlToNewMessage(message_factory, any->type_url());
+    if (typed_message == nullptr) {
+      // If the type URL doesn't correspond to a known proto, give up redacting and treat this
+      // message the same as a `ProtobufWkt::Struct`. See the documented limitation on
+      // `MessageUtil::redact()` for more context.
+      ENVOY_LOG_MISC(warn, "Could not redact ProtobufWkt::Any with unknown type URL {}",
+                     any->type_url());
+      return false;
+    }
+
+    any->UnpackTo(typed_message.get());
+    redact(typed_message.get(), ancestor_is_sensitive);
+    any->PackFrom(*typed_message);
+    return true;
+  }
+  return false;
+}
+
+// To redact a `TypedStruct`, we have to reify it based on its `type_url` to redact it.
+bool redactTypedStruct(Protobuf::Message* message, bool ancestor_is_sensitive) {
+  auto* typed_struct = dynamic_cast<udpa::type::v1::TypedStruct*>(message);
+  if (typed_struct != nullptr) {
+    Protobuf::DynamicMessageFactory message_factory;
+    auto typed_message = typeUrlToNewMessage(message_factory, typed_struct->type_url());
+    if (typed_message == nullptr) {
+      // If the type URL doesn't correspond to a known proto, give up redacting and treat this
+      // message the same as a `ProtobufWkt::Struct`. See the documented limitation on
+      // `MessageUtil::redact()` for more context.
+      ENVOY_LOG_MISC(warn, "Could not redact udpa::type::v1::TypedStruct with unknown type URL {}",
+                     typed_struct->type_url());
+      return false;
+    }
+
+    MessageUtil::jsonConvert(typed_struct->value(), ProtobufMessage::getNullValidationVisitor(),
+                             *typed_message);
+    redact(typed_message.get(), ancestor_is_sensitive);
+    MessageUtil::jsonConvert(*typed_message, *(typed_struct->mutable_value()));
+    return true;
+  }
+  return false;
+}
+
+// `DataSource`, which is annotated as `sensitive` in `TlsCertificate` and `TlsSessionTicketKeys`
+// requires special handling.
+bool redactDataSource(Protobuf::Message* message, bool ancestor_is_sensitive) {
+  auto* data_source = dynamic_cast<envoy::config::core::v3alpha::DataSource*>(message);
+  if (data_source != nullptr) {
+    if (ancestor_is_sensitive) {
+      switch (data_source->specifier_case()) {
+      case envoy::config::core::v3alpha::DataSource::SPECIFIER_NOT_SET:
+        // If the data source is empty, no work is needed.
+        break;
+      case envoy::config::core::v3alpha::DataSource::kFilename:
+        // Don't redact filenames (SecretManagerImplTest::ConfigDumpNotRedactFilenamePrivateKey).
+        break;
+      case envoy::config::core::v3alpha::DataSource::kInlineBytes:
+        // Clear inline bytes and treat it as a string (fall through).
+        data_source->clear_inline_bytes();
```

Comment:
or alternatively can be "[redacted]" corresponding bytes to be set here?
(just wondering if this makes datasource handling easier to follow)

Reply:
I like this idea, although I'd prefer to have the hex value be something human-readable to make it obvious the value is redacted. What do you think about `DEFACED`? htuch@ - wdyt?

Reply:
I don't think there is a major issue just using the ASCII..